File: src/mapBuilder/setValues.go
package main

import (
	"fmt"
	"slices"

	"github.com/utkarsh-naman/3T/src/Constants"
	"github.com/utkarsh-naman/3T/src/model"
)

var gamemap, _ = model.LoadGMap("bin/graph/valued/map1.ttt")
var winStates = getTerminal2(gamemap)
var loseStates []model.State

var workedHistory = make(map[model.State]bool)

func parentFromMap(state model.State) []model.State {
	var parentStates []model.State
	for stateKey := range gamemap {
		if slices.Contains(gamemap[stateKey].NextStates, state) {
			parentStates = append(parentStates, stateKey)
		}
	}
	return parentStates
}

func main() {
	wins()
	model.PrintGMap(gamemap)
	fmt.Println(len(gamemap))
	err := model.SaveGMap(gamemap, "bin/graph/valued/map3.ttt")
	if err != nil {
		return
	}
}

func wins() {
	if len(winStates) == 0 {
		return
	}
	loseStates = loseStates[:0]

	// marking the win states with +inf
	for _, winStateKey := range winStates {
		propsWin := gamemap[winStateKey]
		propsWin.Score = Constants.POSINF
		if len(gamemap[winStateKey].NextStates) != 0 {
			propsWin.WinDepth = maxDepth(winStateKey)
		} else {
			propsWin.WinDepth = 0
		}
		gamemap[winStateKey] = propsWin
	}

	for _, winStateKey := range winStates {
		// working for parent of +inf with -inf
		for _, parentStateKey := range parentFromMap(winStateKey) {
			if !workedHistory[parentStateKey] {
				parentProps := gamemap[parentStateKey]
				parentProps.Score = Constants.NEGINF
				gamemap[parentStateKey] = parentProps
				if len(parentProps.NextStates) > 0 {
					parentProps.LoseDepth = minDepth(parentStateKey) + 1
					gamemap[parentStateKey] = parentProps
				}
				loseStates = append(loseStates, parentStateKey)
				workedHistory[winStateKey] = true
			}
		}
	}
	loses()
	return
}

func loses() {
	if len(loseStates) == 0 {
		return
	}
	winStates = winStates[:0]
	for _, loseStateKey := range loseStates {
		for _, parent := range parentFromMap(loseStateKey) {
			if !workedHistory[parent] {
				if isAllNeg(gamemap[parent].NextStates) {
					winStates = append(winStates, parent)
				} else {
					if gamemap[parent].Score == 0 {
						parentProps := gamemap[parent]
						parentProps.Score = zeroScoreReset(parent)
						gamemap[parent] = parentProps
					}
				}
			}
		}
	}
	wins()
	return
}

func minDepth(state model.State) int8 {
	var mind int8 = 10
	for _, childStateKey := range gamemap[state].NextStates {
		mind = min(mind, gamemap[childStateKey].WinDepth)
	}
	return mind
}

func maxDepth(state model.State) int8 {
	var maxd int8 = 0
	for _, childStateKey := range gamemap[state].NextStates {
		maxd = max(maxd, gamemap[childStateKey].LoseDepth)
	}
	return maxd
}

func isAllNeg(children []model.State) bool {
	for _, child := range children {
		if gamemap[child].Score != Constants.NEGINF {
			return false
		}
	}
	return true
}

func getTerminal2(gamemap model.GMap) (terminalWinStates []model.State) {
	var terminalWinMask model.State = 1 << 31
	var terminalTieMask model.State = 1 << 30

	for stateKey := range gamemap {
		if stateKey&terminalWinMask == 0 { // game over
			if stateKey&terminalTieMask == 0 { // game over & not a tie, therefore def a win
				terminalWinStates = append(terminalWinStates, stateKey)
			}
		}
	}
	return
}

func zeroScoreReset(parent model.State) float32 {
	var score float32 = 0
	var children []model.State = gamemap[parent].NextStates
	//fmt.Println("for parent: ", parent)
	for _, childStateKey := range children {
		//fmt.Print("\tchild state:\t", childStateKey)
		childScore := gamemap[childStateKey].Score
		//fmt.Print("\tscore:", childScore, "\n")
		if childScore == Constants.NEGINF {
			score += 1
		} else if childScore == Constants.POSINF {
			score += -1
		} else if childScore > 0 {
			score += childScore
		}
	}
	score = score / float32(len(children))
	return score
}

File: Rust/src/mapbuilder/set_values.rs
use std::cmp::{max, min};
use std::collections::HashMap;
use crate::constants::values::{NEGINF, POSINF};
use crate::mapbuilder::set_terminal_values::get_terminal_values;
use crate::model::map::GameMap;
use crate::model::state_model::State;
use crate::utils::prev_states::parent_of;

pub fn set_values(mut map: GameMap) ->GameMap{
    let win_states = get_terminal_values(&map);
    let mut worked_history: HashMap<State, bool> = HashMap::new();

    wins(&win_states, &mut worked_history, &mut map);
    map
}


fn wins(win_states: &Vec<State>, worked_history : &mut HashMap<State, bool>, map: &mut GameMap) {
    if win_states.len() == 0 {
        return;
    }
    let mut lose_states: Vec<State> = vec![];



    for winkey in win_states{

        let win_wd = if map[winkey].next_state.is_empty() {
                0
            } else {
                max_depth(map, winkey)
            };



        if let Some(win_props) = map.get_mut(winkey) {
            win_props.score = POSINF;
            win_props.win_depth = win_wd;
        }

        for parent_key in parent_of(&map, winkey){
            if worked_history.insert(parent_key, true).is_none() {


                let max_ld = if map[&parent_key].next_state.len() > 0 {
                        min_depth(map, &parent_key) + 1
                    } else {
                        0 // Default or handle as needed, though logic suggests > 0
                    };

                if let Some(parent_props) = map.get_mut(&parent_key) {
                    parent_props.score = NEGINF;
                    if parent_props.next_state.len() > 0 {
                        parent_props.lose_depth = max_ld;
                    }
                    lose_states.push(parent_key);
                }
            }
        }
    }

    loses(&lose_states, worked_history, map);
    return
}



fn loses(lose_states: &Vec<State>, worked_history: &mut HashMap<State, bool>, map: &mut GameMap) {
    if lose_states.len() == 0 {
        return;
    }
    let mut win_states: Vec<State> = vec![];

    for lose_key in lose_states{
        for parent_key in parent_of(&map, lose_key){
            if !worked_history.insert(parent_key, true).is_none() {
                if is_all_neg(&map, &map[&parent_key].next_state) {
                    win_states.push(parent_key);
                } else{
                    if map[&parent_key].score == 0.0{
                        let parent_new_score = zero_score(map, &parent_key);
                        if let Some(parent_props) = map.get_mut(&parent_key){
                            parent_props.score = parent_new_score;
                        }
                    }
                }
            }
        }

    }



    wins(&win_states, worked_history, map);
    return
}

fn min_depth(map: &GameMap, state: &State)->u8{
    let mut mind = 10;
    for &child in &map[state].next_state {
        mind = min(mind, map[&child].win_depth);
    }
    mind
}

fn max_depth(map: &GameMap, state: &State)->u8{
    let mut maxd = 0;
    for &child in &map[state].next_state {
        maxd = max(maxd, map[&child].lose_depth);
    }
    maxd
}

fn is_all_neg(map: &GameMap, children : &Vec<State>) -> bool {
    for child in children {
        if map[child].score != NEGINF{
            return false
        }
    }
    true
}

fn zero_score(map: &GameMap, state: &State)->f32{
    let mut score:f32 = 0.0;
    let children = &map[state].next_state;

    for child in children{
        let child_score = map[child].score;
        if child_score == NEGINF{
            score += 1.0;
        } else if child_score == POSINF{
            score -=1.0;
        } else{
            score -= child_score;
        }
    }
    score = score/(children.len() as f32);
    score
}